### PR TITLE
[IMAGING-340] Support Extensions to the PNG 1.2 Specification, Version 1.5.0

### PR DESCRIPTION
**Pull request by Glavo**
I hope to support [Extensions to the PNG 1.2 Specification, Version 1.5.0](http://ftp-osl.osuosl.org/pub/libpng/documents/pngext-1.5.0.html) in this PR and be ready to support the APNG file format in subsequent PR.

Completed works:

* Add an enum representing the PNG extension type;
* Refactoring `PngImageParser::readChunks` and `ChunkType` to improve maintainability makes it easier to add new Chunk types later;
* Support reading EXIF from PNG images.
* Merge EXIF into PNG metadata;

### CODE REVIEW

**Comment by garydgregory** [comment]
FYI, this will not be considered without unit tests.

**Comment by Glavo** [comment]
> FYI, this will not be considered without unit tests.

Before that, I need to find a way to generate test data.
The new extension lacks tool support. I think I may need to use libpng directly to generate test data.

**Comment by Glavo** [comment]
When I created the test, I noticed the type of `EXIF_TAG_EXIF_IMAGE_WIDTH` and `EXIF_TAG_EXIF_IMAGE_LENGTH` is `TagInfoShort`.

However, in the [standard document](https://www.cipa.jp/std/documents/e/DC-X008-Translation-2019-E.pdf), the field type is SHORT or LONG.

Unfortunately, the test image I found uses the LONG type field to record `ExifImageWidth` and `ExifImageLength`, so I found this problem.

Should I solve this problem in this PR? Or do I need to open a new PR?

**Comment by Glavo** [comment]
I updated this PR, followed the requirements of IMAGING-341, and recorded the standard version in the document.

Can someone review this PR?

File: src/main/java/org/apache/commons/imaging/formats/png/package-info.java
Comment (kinow) [inline src/main/java/org/apache/commons/imaging/formats/png/package-info.java:26]: Thank you! :clap:

File: src/main/java/org/apache/commons/imaging/formats/png/PngImageMetadata.java
Comment (kinow) [inline src/main/java/org/apache/commons/imaging/formats/png/PngImageMetadata.java:27]: Can you add javadocs to the new public classes/methods/fields/etc, please? Doesn't need to be long, but we need a `@since $current-snapshot-version-minus-snapshot` here and in the public members added to existing code.

File: src/main/java/org/apache/commons/imaging/formats/png/PngImageParser.java
Comment (kinow) [inline src/main/java/org/apache/commons/imaging/formats/png/PngImageParser.java:193]: :ok_man: :clap:  bravo, @Glavo

File: src/main/java/org/apache/commons/imaging/formats/png/PngImageParser.java
Comment (kinow) [inline src/main/java/org/apache/commons/imaging/formats/png/PngImageParser.java:285]: I think we should either log and/or raise an error for any other type here.

File: src/main/java/org/apache/commons/imaging/formats/png/PngImageParser.java
Comment (kinow) [inline src/main/java/org/apache/commons/imaging/formats/png/PngImageParser.java:305]: Docs for new public members with the `@since` tag, please :+1:

**Review by kinow — COMMENTED** [review]
Had a initial look @Glavo , I left a few comments, if you could reply/update it, I'll have a final look with the IDE and then review/merge it :+1: Thanks!

**Comment by Glavo** [comment]
The javadoc has been updated to add the `@since` tag.

**Comment by StefanOltmann** [comment]
@Glavo Thank you!

I don’t know if this PR will ever be pulled, because by now I think that this project is pretty much dead.

But you did not waste your time on this, because I pulled it into my Kotlin port [Ashampoo Kim](https://github.com/ashampoo/kim). So you may be glad to hear that your code is used in production (for example by [Ashampoo Photos](https://www.ashampoo.com/photos)) and people benefit from it. 🙂

I invite you to contribute directly to my fork in the future.

**Comment by kinow** [comment]
> I don’t know if this PR will ever be pulled, because by now I think that this project is pretty much dead.

Certainly not dead. Maintained by volunteers, yes. I am busy with deadlines at work, but watching and planning  a future development cycle, fix security bugs (a lot happens there without activity on GitHub for a reason, so again, not dead).

I think it is good to have more projects like what you created to give more options to users.

But leave a comment saying a project is dead and promote your own tool doesn't look very good, IMHO.

And apologies again for the radio silence, just busy with $work and not able to volunteer time at this point (happy if any other committer takes over, otherwise March I should be available).

Cheers

**Comment by StefanOltmann** [comment]
@kinow 

Thank you for reaching out and providing an update.

I must admit, I've encountered some frustration recently. Over the past few months, the only noticeable activity on the project seems to be from Gary, who has been addressing minor style issues but not significantly contributing otherwise. It appears he hasn't been thoroughly reviewing code in pull requests and may not grasp the project's domain.

Even without unit tests, confirming the correctness of the bugfix in my pull request or validating the provided test should have been straightforward for someone familiar with the codebase and its intricacies. This is the kind of scrutiny and understanding that someone actively engaged in the project could provide.

I recall observing this level of engagement from you some time ago, which is why I directed my concerns to you.

Maybe the project isn’t dead, given the abundance of open issues and pull requests coupled with the lack of substantial development, it's difficult to characterize it as an actively progressing endeavor.

I am glad to hear that you plan to pick up the work here, because this project deserves real maintenance.

How Gary acts on my PRs is just ridiculous. First he wanted me to include a test image and provide a test - both would not have been necessary. After I did go the extra mile he doesn’t want the test (because he does not seem to understand what it does) and now all of the sudden the inclusion of the test image is a problem. That’s really sad and ridiculous maintenance. I believe this projects deserves better.

**Comment by garydgregory** [comment]
This component is definitely not dead.

**Comment by BuZZ-dEE** [comment]
@kinow @garydgregory @Glavo What is the progress of this pull request?

**Comment by garydgregory** [comment]
> @kinow @garydgregory @Glavo What is the progress of this pull request?

Two items I can see at a glance: Address the review comments and rebase on git master.

**Comment by Glavo** [comment]
> What is the progress of this pull request?

@BuZZ-dEE I will pick up this PR again after some time. But before that, I hope someone can help me solve the problem mentioned before: 

https://github.com/apache/commons-imaging/pull/269#issuecomment-1381983655

> When I created the test, I noticed the type of `EXIF_TAG_EXIF_IMAGE_WIDTH` and `EXIF_TAG_EXIF_IMAGE_LENGTH` is `TagInfoShort`.
> 
> However, in the [standard document](https://www.cipa.jp/std/documents/e/DC-X008-Translation-2019-E.pdf), the field type is SHORT or LONG.
> 
> Unfortunately, the test image I found uses the LONG type field to record `ExifImageWidth` and `ExifImageLength`, so I found this problem.

@garydgregory @kinow I haven't had time to look into this lately, do you have time to look at it soon?

**Comment by kinow** [comment]
Some Git confligts, so I've squashed the 11 commits into a single one to simplify rebasing it.

File: src/main/java/org/apache/commons/imaging/formats/png/Extension.java
Comment (kinow) [inline src/main/java/org/apache/commons/imaging/formats/png/Extension.java:24]: I wonder if this should be called `PngExtension`, as there are `PngImageMetadata`, `PngImageParser`, etc. inside the `oac.imaging.formats.png` package. Although we also have `ChunkType` :shrug: , I think that's fine.

File: src/main/java/org/apache/commons/imaging/formats/png/ChunkType.java
Comment (kinow) [inline src/main/java/org/apache/commons/imaging/formats/png/ChunkType.java:206]: Good simplification here :+1:

File: src/conf/spotbugs-exclude-filter.xml
Comment (kinow) [inline src/conf/spotbugs-exclude-filter.xml:185]: These look good to me!

File: src/main/java/org/apache/commons/imaging/formats/png/PngImageParser.java
Comment (kinow) [inline src/main/java/org/apache/commons/imaging/formats/png/PngImageParser.java:529]: Looks good!  :+1: 

image

**Review by kinow — COMMENTED** [review]
Code looks good, adding the new chunk types, and also some code simplification (very welcome!). The spotbugs exclusions look good. I just want to spend a bit more of time now reading the spec linked (thanks for that, BTW!). Should approve & merge between today and tomorrow as I am on long holidays now, finally. Sorry the delay!

**Review by kinow — APPROVED** [review]
The extensions doc is quite simple, describing the chunks we added, the one deprecated (gift, which is not handled here, and this is recommended in the spec), and also the ones that are not described in the doc that we also do not handle.

The part about int/float shouldn't be a problem as I think we simply read the bytes as-they-are.

Thanks for the PR @Glavo ! Approving, will update changes.xml, merge it, and close the issue linked. Thanks!

**Pull request merged**
